### PR TITLE
BZ1333939: disable custom route hosts and certificates

### PR DESCRIPTION
**Pull request by Miciah**
#### router: Move `RejectionRecorder` to its own file

Move the `RejectionRecorder` interface and default implementation from `unique_host.go` into its own file, `rejection_recorder.go`.
#### router: Factor `RouteValidator` out of `UniqueHost`

Define a new plugin wrapper, `RouteValidator`.  Move existing validation from `UniqueHost` into `RouteValidator`:

• Reject the route if it is not from an allowed namespace.

• Override the hostname using the defined template if it is blank or if the plugin wrapper is configured to override hostnames.

Move `allowedNamespaces` and `hostForRoute` from `UniqueHost` to `RouteValidator`.

Move `RouteSelectionFunc` from `pkg/cmd/infra/router/router.go` to `pkg/router/controller/route_validator.go`, and modify it to take the hostname template and hostname-override flag as parameters instead of taking `RouterSelection` as a receiver and reading the parameters from there.

Call `RouteSelectionFunc` from `NewRouteValidator`, which now sets the `hostForRoute` callback, instead of calling `RouteSelectionFunc` in `NewUniqueHost`'s caller.

Move namespace validation, host validation, and host template expansion from `UniqueHost`'s `HandleEndpoints`, `HandleRoute`, and `HandleNamespace` callbacks to `RouteValidator`'s.

Update callers of `NewUniqueHost` to drop the `hostToRoute` argument if they do not need route validation or to use `NewRouteValidator` to create a `RouteValidator` that wraps a `UniqueHost` if they do need route validation.
#### router: add `--override-exceptions` flag

Add `--force-exceptions` flag for `oadm router` and `--override-exceptions` flag and `ROUTER_OVERRIDE_EXCEPTIONS` environment variable for the router itself.

Add `ForceExceptions` string array field to `RouterConfig` and update `NewCmdRouter` and `RunCmdRouter` accordingly.

Add `OverrideExceptions` string array field to `RouterSelection` and update the `RouterSelection` `Bind` and `Complete` methods accordingly.

Add the exceptions parameter to `NewRouteValidator` and `routeSelectionFunc` (the former passing it to the latter).

Update callers of `NewRouteValidator` to pass in the exceptions parameter.

Modify `routeSelectionFunc` to check the exceptions before overriding a route's host and leave it as is if the route's namespace is in the list of exceptions.

Add `TestOverrideExceptions`.

Regenerate completions.
#### router: add `--allow-custom-certificates` flag

Add an `--allow-custom-certificates` flag (default true) for `oadm router` and for the router, and `ROUTE_ALLOW_CUSTOM_CERTIFICATES` environment variable for the router, that specify whether routes that specify custom certificates will be accepted.

Namespaces specified with `--force-exceptions`/`--override-exceptions` will still be allowed to specify custom certificates even if `--allow-custom-certificates=false` is specified.

Add `AllowCustomCertificates` Boolean to `RouterConfig` and update `NewCmdRouter` and `RunCmdRouter` accordingly.

Add `AllowCustomCertificates` Boolean to `RouterSelection` and update the `RouterSelection` `Bind` and `Complete` methods accordingly.

Add `allowCustomCertificates` and `overrideExceptions` to `RouteValidator`.

Add a new parameter for the flag to `NewRouteValidator`, and update `NewRouteValidator` to set the new fields in `RouteValidator`.

Update `RouteValidator`'s `HandleRoute` method to reject any route with a custom certificates if the route's namespace is not in the list of exceptions and the new flag is set.

Update callers of `NewRouteValidator` to pass in the new flag.

Add `TestAllowCustomCertificates`.

Regenerate completions.
## 

openshift-bot, please [test]!

### CODE REVIEW

**Comment by Miciah** [comment]
I regenerated completions to address test_pull_requests_origin_check failure.  I am not sure why `test-end-to-end-docker.sh` failed—it does not obviously look like the failure was caused by the changes in this PR.

**Comment by Miciah** [comment]
I am not sure what is causing the route validation failure in the end-to-end tests, and we are having difficulties with AWS right now, but we are trying to get this in 3.2.1.  @ramr, would you be able to give this at least a high-level review so we can address any concerns over the general approach or naming issues as soon as possible?

**Comment by Miciah** [comment]
I could not reproduce the test failure in a devenv (the "Validating routed app response" step passed).  Please [test] again!

**Comment by Miciah** [comment]
This time it failed on TestImageStreamImport, which I have seen on other PRs and which looks unrelated to my changes.

@knobunc, we're trying to get this in for 3.2.1.  Would you be able to review this PR, or help us find someone on the networking team who can review it ASAP?

**Comment by knobunc** [comment]
LGTM [test]

@smarterclayton can you take a look at this and see if this is what you had in mind?

**Comment by smarterclayton** [comment]
So I had not anticipated making this a feature.  My though was just a custom template that didn't use certs.

**Comment by smarterclayton** [comment]
This change as is way too much risk for 3.2 - I would be concerned about this even in master.

**Comment by knobunc** [comment]
@smarterclayton: So just a custom router image with that stuff cut out?

**Comment by liggitt** [comment]
not even a custom image... just a custom template

**Comment by Miciah** [comment]
We want to allow custom certificates for routes in privileged namespaces (e.g., for the registry).

**Comment by smarterclayton** [comment]
Add an exception in the template.

**Comment by Miciah** [comment]
Rebased.

**Comment by knobunc** [comment]
Seems reasonable.  @ramr @rajatchopra PTAL

**Comment by Miciah** [comment]
I messed up the rebase.  Shouldn't take long to fix...

**Comment by smarterclayton** [comment]
Do we need this now?

**Comment by Miciah** [comment]
We would very much like to have it for Online so that we do not need to maintain a custom template.  Using a custom template means we must rebase it to get upstream template changes, we must change the template in order to change the namespace whitelist, and we cannot prevent custom certificates from being copied into HAProxy's configuration directory (we can only omit the frontend configuration, not the certificates themselves), which poses performance and security concerns.

**Comment by Miciah** [comment]
I fixed the rebase and verified that `make clean all test-unit WHAT=pkg/router/controller` passes unit tests with each commit.

File: pkg/router/controller/route_validator.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+		return strings.Trim(s, "\"'")
+	}
+}
+
+// RouteValidator implements the router.Plugin interface and validates
+// routes before passing them along to another router.Plugin.
+type RouteValidator struct {
+	plugin router.Plugin
+
+	recorder RejectionRecorder
```
Comment (rajatchopra) [inline pkg/router/controller/route_validator.go]: Why do we need an entire new object for just logging? Are we expected to persist this information ever?

Comment (Miciah) [inline pkg/router/controller/route_validator.go]: What do you mean? Outside of tests, all users of `RouteValidator` are creating a recorder anyway for other plugins, and all the plugins in the chain (including the `RouteValidator`) use the same recorder object.

**Comment by brenton** [comment]
@abhgupta, @Miciah, we're past dcut for 3.3. Can this simply ship with 3.4 or do we need it merged now?  If so we need understand how risky this change is.

cc @rajatchopra

**Comment by abhgupta** [comment]
@brenton @Miciah we would have liked to get this into 3.3 as it would have allowed us to get rid of the custom router image that we are currently using. However, we can wait for this to be properly reviewed and included in 3.4

**Comment by smarterclayton** [comment]
At this point why aren't we spending this effort on making custom
certs safe?  Is that something Miciah can focus on?

**Comment by abhgupta** [comment]
@smarterclayton The work for this feature is already done and the PR has gone through some review as well. At this point, I would like to get this into 3.4 and then we can start the work for validating custom certs.

@Miciah can you please rebase this PR to resolve the conflicts?

**Comment by smarterclayton** [comment]
I am concerned that my feedback has consistently to push back and say that we should fix the underlying problem (for security of certs) rather than band-aid around it.  Given the amount of work invested in this PR so far, I'm also concerned we could have fixed that problem by now.  Instead, we have no fix, but a fix to a fix.  

Just because this PR has gone through review doesn't impact that it isn't helping to close the underlying problem.

**Comment by smarterclayton** [comment]
Can you make a case that either of these proposed changes is actually useful to a majority of admins?  Because making custom certs safe is of interest to 100% of cluster admins.

**Comment by abhgupta** [comment]
@smarterclayton The only use case that I have is that we might want to block V3 Online free-tier users from creating custom routes. This was the case with V2 Online free tier. Beyond that, this PR was mostly to move us away from using a custom router image in Online.

**Comment by liggitt** [comment]
I would expect us to limit custom cert usage via the API and quota objects for per-user limits, not in the router. 

I would also like to see online stop using a custom router image, but I see that being accomplished by fixing the underlying issues:

* route claiming (part API, part router)
* cert/key validation (API level and/or router level)
* correct mapping of custom certs to scope to the route they came from (instead of default SNI behavior)

**Comment by smarterclayton** [comment]
Route claiming is broken in online?  I don't know how.

Otherwise agree - all of these are API level restrictions (although sharded
routers is a use case where the restriction may differ by router).

On Aug 22, 2016, at 7:43 PM, Jordan Liggitt <notifications@github.com>
wrote:

I would expect us to limit custom cert usage via the API and quota objects
for per-user limits, not in the router.

I would also like to see online stop using a custom router image, but I see
that being accomplished by fixing the underlying issues:

   - route claiming (part API, part router)
   - cert/key validation (API level and/or router level)
   - correct mapping of custom certs to scope to the route they came from
   (instead of default SNI behavior)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9043#issuecomment-241584756>, or mute
the thread

**Comment by liggitt** [comment]
> Route claiming is broken in online?  I don't know how.

Ah, right. creationTimestamp plus immutable route.spec.host makes for deterministic claiming

**Comment by Miciah** [comment]
> limit custom cert usage via the API and quota objects for per-user limits

Do you mean a limit on the number of Route objects, or are you thinking of something that could limit certificates specifically? Do we already have GitHub issues or Trello cards for either of these?

> - route claiming (part API, part router)

[UniqueHost ensures malicious users cannot usurp hosts](https://github.com/openshift/origin/blob/9f6ce76aa86ea6cb31f975387db6fa6dd8280be1/pkg/router/controller/unique_host.go#L140-L145) (from either infrastructure components' or other users' routes).

> - cert/key validation (API level and/or router level)

Right now, most egregiously, [ExtendedValidator's host validation is disabled](https://github.com/openshift/origin/blob/9f6ce76aa86ea6cb31f975387db6fa6dd8280be1/pkg/route/api/validation/validation.go#L117-L121), which means we allow certificates for which the CN does not match the route's host.

Supposing we enabled host validation, it is important to note that extended validation is performed using the Route object and not the host that the template writes into `haproxy.config`, so if we continue to disable custom hosts by using a custom template to override the host specified in the Route, then the user could still specify a Route with a bogus host and a matching bogus certificate, which would mean enabling host validation would not necessarily protect against usurpation.

However, if we enabled host validation and convinced ourselves that our validation were perfect, then there would be less need to disable custom hosts and certificates in the first place (at least outside the Online use-case).  So if we decide to allow custom hosts and drop the custom template, enable extended route validation in configuration, and enable ExtendedValidator's host validation in the code, then UniqueHost's claim protection together with ExtendedValidator's host validation should protect against usurpation by either host or certificate.

Besides usurpation, the other major concern is DoSing the router with invalid configuration.  One concern about extended validation is that it loads and parses certificates using Go libraries.  Go's implementation of X.509 etc. might be perfect, or it might not be, but what really matters is that Go's implementation accept a subset of the inputs that HAProxy's implementation accepts (and, conversely, that Go reject anything that HAProxy rejects—keeping in mind that HAProxy's implementation of X.509 etc. likewise might be perfect or might not be).  If there are edge cases where a certificate is valid and Go accepts it but HAProxy still rejects it, that is a potential DoS.

To mitigate risk, would it be reasonable to keep the old `haproxy.config` when rewriting a new from from the template and roll back to the old version if the new one causes haproxy to fail to start? This approach would be far from ideal—one bad route could cause the rollback of many route updates, and we would not necessarily be able to identify the culpable route automatically—but it would somewhat mitigate the risk of DoSing existing routes.

Final concerns are around attack surface and performance.  The former is abstract until an exploit is found, but the latter can be quantified by creating a few thousand routes with custom certificates and timing or profiling the router's startup.  Does that sound reasonable? (Has anyone already done this?)

> - correct mapping of custom certs to scope to the route they came from (instead of default SNI behavior)

Can you elaborate on this?

**Comment by liggitt** [comment]
> Right now, most egregiously, ExtendedValidator's host validation is disabled, which means we allow certificates for which the CN does not match the route's host.

Making sure the cert CN matches the route host is insufficient, since certs can also contain SANs. Rather than requiring the cert's CN to match the route's host, we should instead tell haproxy explicitly which cert to use for which host.

>> correct mapping of custom certs to scope to the route they came from (instead of default SNI behavior)
>
> Can you elaborate on this?

That is the `host<->cert` mapping I was referring to, rather than relying on haproxy to use SNI and introspection of all the certs' CNs/SANs to pick a cert. It looks like there is a way to specify this mapping to haproxy:

https://coolaj86.com/articles/haproxy-with-https-tls-ssl/
http://marc.info/?l=haproxy&m=143986084312117&w=2
http://cbonte.github.io/haproxy-dconv/1.6/configuration.html#5.1-crt-list

**Pull request closed without merging**